src/treq/client.py

```diff
@@ -255,11 +236,99 @@ def gotResult(result):
+            )
+
         return d.addCallback(_Response, cookies)
 
+    def _request_body(self, data, files, json, stacklevel):
+        """
+        Here we choose a right producer based on the parameters passed in.
+
+        :params data:
+            Arbitrary request body data.
+
+            If *files* is also passed this must be a :class:`dict`,
+            a :class:`tuple` or :class:`list` of field tuples as accepted by
+            :class:`MultiPartProducer`. The request is assigned a Content-Type
+            of ``multipart/form-data``.
+
+            If a :class:`dict`, :class:`list`, or :class:`tuple` it is
+            URL-encoded and the request assigned a Content-Type of
+            ``application/x-www-form-urlencoded``.
+
+            Otherwise, any non-``None`` value is passed to the client's
+            *data_to_body_producer* callable (by default,
+            :class:`IBodyProducer`), which accepts file-like objects.
+
+        :params files:
+            Files to include in the request body, in any of the several formats
+            described in :func:`_convert_files()`.
+
+        :params json:
+            JSON-encodable data, or the sentinel `_NOTHING`. The sentinel is
+            necessary because ``None`` is a valid JSON value.
+        """
+        if json is not _NOTHING and (files or data):
+            warnings.warn(
+                (
+                    "Argument 'json' will be ignored because '{}' was also passed."
+                    " This will raise TypeError in the next treq release."
+                ).format("data" if data else "files"),
+                DeprecationWarning,
+                stacklevel=stacklevel + 1,
+            )
+
+        if files:
+            # If the files keyword is present we will issue a
+            # multipart/form-data request as it suits better for cases
+            # with files and/or large objects.
+            files = list(_convert_files(files))
+            boundary = str(uuid.uuid4()).encode('ascii')
+            if data:
+                data = _convert_params(data)
+            else:
+                data = []
+
+            return (
+                multipart.MultiPartProducer(data + files, boundary=boundary),
+                b'multipart/form-data; boundary=' + boundary,
+            )
+
+        # Otherwise stick to x-www-form-urlencoded format
+        # as it's generally faster for smaller requests.
+        if isinstance(data, (dict, list, tuple)):
+            return (
+                self._data_to_body_producer(urlencode(data, doseq=True)),
+                b'application/x-www-form-urlencoded',
+            )
+        elif data:
+            return (
+                self._data_to_body_producer(data),
+                None,
+            )
+
+        if json is not _NOTHING:
+            return (
+                self._data_to_body_producer(
+                    json_dumps(json, separators=(u',', u':')).encode('utf-8'),
```

Comment:
The `u` strings shouldn't be needed right? They would only make the output unicode in python2, but the `encode` makes it bytes anyways.

When will treq drop python 2 anyways?

Reply:
I suppose the `u` isn't necessary, but I will leave it as-is since I was only moving it.

Treq will probably drop Python 2 once the Twisted release is out. I want to kick out a release with these deprecations first, though.